---
title: "Poisson_fSuSiE"
author: "William R.P. Denault"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Introduction to Poisson-fSuSiE}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r , include=FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.width = 5,
  fig.height = 3,
  fig.align = "center",
  fig.cap = "&nbsp;",
  dpi = 175
  )
```


```{r setup, include=FALSE}
rm(list = ls()) 
library(fsusieR)
library(susieR)
library(ebnm)
data("poisson_fsusie_e2")
```

These data are based on sample real GTeX RNAseq counts where we model the effect of 
a SNP on these count using Binomial thinning.

Those are simulated data of structured counts $Y$ each row being the observed "profile"
for an individual and in $X$ we store the genotype data, among those the  effect   SNP  is

```{r}
true_pos=example$true_pos
true_pos
```
Here a typical count profile

```{r}
plot(example$Y[1,], main="RNAseq count for individual 1 over the region of interest")
```



The effect of the SNP  on the count is

```{r}
 
plot(example$lf[[1]] , type="l", main = "effect of the causal SNP") 
X=example$X
Y=example$Y
 
```


A simple to handle count for fine mapping is to use $log(x+1)$ transform or the Haar-Fisz transform 
when dealing with structured sequence of count as we are doing

```{r}

res0 = susiF(X=X, Y=log1p(Y),L=3)
res1= susiF(X=X,Y=HFT(Y),L=3)
```

Unfortunately those results are not quite correct

```{r}
res0$cs
res1$cs
true_pos
```



 

```{r}

res_poisF = Pois_fSuSiE(Y=Y,X=X ,L=3 , max.iter=2)

```
 It contains a single CS with lead SNP being the causal SNPS
 
```{r}
res_poisF$susiF.obj$cs
```
 
 
 
```{r}
plot(res_poisF$susiF.obj$fitted_func[[1]])
lines(example$lf )
```

 
  
