---
title: "Introduction to fSuSiE"
author: "William R.P. Denault"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: no
    highlight: textmate
    theme: readable
vignette: >
  %\VignetteIndexEntry{Introduction to fSuSiE}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r knitr-opts, include=FALSE}
knitr::opts_chunk$set(collapse = TRUE,comment = "#>",fig.width = 5,
                      fig.height = 3,fig.align = "center",
					  fig.cap = "&nbsp;",dpi = 120)
```

## Overview

fSuSiE implements a Bayesian variable selection method based on the
"Functional Sum of Single Effects" (fSuSiE) model.  The fSuSiE model
extends the SuSiE model to functional traits. fSuSiE (and SuSiE) are
particularly well suited to settings where some of the variables are
strongly correlated and the true effects on the trait are very
sparse. This is exactly the situation we face in genetic
fine-mapping. The functional traits of interest are molecular
measurements from sequencing assays (e.g., ChIP-seq, ATAC-seq), and
DNA methylation. However, the methods are quite general and could be
useful for other problems inside and outside genetics.

## The fSuSiE model

fSuSiE fits the linear regression model
$$
{\bf Y} = {\bf X} {\bf B} + {\bf E},
$$

where $\bf Y$ is an N x T matrix containing the $N$ trait observations
at $T$ locations; $\bf X$ is the N x p matrix of single nucleotide
polymorphisms (SNP) in the region to be fine-mapped; $\bf B$ is a p x
T matrix whose element $b_{jt}$ is the effect of SNP $j$ on the trait
at location t; and $\bf E$ is a matrix of error terms.

Analogous to regular fine-mapping, the matrix $\bf B$ is expected to
be row-wise sparse, that is, most of the rows of $\bf B$ will be
zero. (Non-zero rows correspond to SNPs that affect the trait.) The
goal, in short, is to identify the non-zero rows of ${\bf B}$. To
accomplish this task, we extend SuSiE to functional traits: we express
${\bf B}$ as a sum of "single effects",
$$
{\bf B} = \sum_{l=1}^L {B}^{(l)},
$$
where each single-effect matrix ${B}^{(l)}$ has only one non-zero row,
corresponding to a single causal SNP. We have developed a fast
procedure for fitting this model by (i) projecting the data in the
wavelet space, then (ii) using a version of the iterative Bayesian
stepwise selection (IBSS) algorithm adapted for fSuSiE.

fSuSiE also outputs Credible Sets (CSs). Each CS is defined so that it
has a high probability of containing a variable with a non-zero effect
while at the same time being as small as possible.

We illustrate these ideas in the examples below.

## The example fine-mapping data set

```{r load-packages}
library(susieR)
library(fsusieR)
library(ggplot2)
library(cowplot)
```

We illustrate fSuSiE on a simulated molecular trait in which two of
the candidate genetic variants (SNPs) affect the trait. The molecular
trait spans a region of 120 locations.

First, we simulate the SNP effects (blue and green in the plot):

```{r sim-effects, fig.height=2, fig.width=5}
set.seed(1)
f1 <- simu_IBSS_per_level(7)$sim_func
f2 <- simu_IBSS_per_level(7)$sim_func
par(mar = c(2,2,2,2))
plot(f1,type = "l",col = "royalblue",lwd = 2)
abline(a = 0,b = 0,lty = "dotted")
lines(f2,type = "l",col = "limegreen",lwd = 2)
```

Notice that each of the SNPs increases or decreases at many locations,
but in a mostly spatially smooth way.

Now we simulate the molecular trait. To do this, we use some genotype
data included in susieR. 

```{r sim-trait}
data(N3finemapping)
rsnr  <- 0.5
pos1  <- 25
pos2  <- 75
X     <- N3finemapping$X[,1:100]
n     <- nrow(X)
nloc  <- length(f1)
noise <- matrix(rnorm(n*nloc,sd = var(f1)/rsnr),n,nloc)
Y     <- tcrossprod(X[,pos1],f1) + tcrossprod(X[,pos2],f2) + noise
```

The end result is a fine-mapping data set containing the genotype data
at 100 SNPs (stored as a 574 x 100 matrix ${\bf X}$ ) and the
molecular trait data at 128 locations (stored as a 574 x 128 matrix
${\bf Y}$).

Note that the true causal SNPs are at SNP positions 25 and 75
(columns 25 and 75 in the X matrix).

### Run fSuSiE

Running fSuSiE requires a Y matrix in which each line corresponds to
an individual profile/time series. X the genotype matrix in which each
line represents the individual genotype and the column the different
SNPs. L is an upper bound on the number of causal SNPs within the
regions (we advise keeping this parameter below 20).

```{r run-susiF, results="hide"}
out <- susiF(Y,X,L = 10)
```

FSuSiE automatically selects the number of causal effects (here
two). The effects detected are summarized in terms of credible
sets. Each credible set contains the likely causal covariates for a
given effect. You can access them simply as
 
```{r}
out$cs
```
As you can see, the two CSs contain the correct causal SNPs.

### Visualization

The easiest way to visualize the result is to use the
\code{plot_susiF} function.

```{r}
plot_susiF(out )
```

You can also access the information directly in the output of susiF in
the fitted_func part of the output. See below
 
```{r} 
plot( f1, type="l", main="Estimated effect 1", xlab="")
lines(get_fitted_effect(out,l=2),col='blue' )
abline(a=0,b=0)
legend(x= 35,
       y=3,
       lty= rep(1,2),
       legend = c("effect 1"," fSuSiE est "),
       col=c("black","blue" )
)
plot( f2, type="l", main="Estimated effect 2", xlab="")
lines(get_fitted_effect(out,l=1),col='green' )
abline(a=0,b=0)
legend(x= 20,
       y=-1.5,
       lty= rep(1,2),
       legend = c("effect 2"," fSuSiE est "),
       col=c("black","green" )
)
```

#### Posterior inclusion probabilities

You can also access the Posterior inclusion probabilities directly
from the fitted object. Similarly, you can specify the plot_susiF
function to display only the PiP plot.
 
```{r}
 
out$pip[1:10]

plot_susiF(out, pip_only=TRUE)
```
 
 
 
 
#### Extracting the regions affected by the credible sets
 We provide a build-in function to retrieve the regions where the credible bands are "crossing zero"/i.e., the effects are likely not to be 0 in this region. 
```{r}
 affected_reg( out)
  
```


### Prediction

You can also check the predictive power of susiF by checking the fitted from the output susiF

 
```{r}
f1_obs <- f1
f2_obs <- f2
true_sig  <- matrix( X[ ,pos1], ncol=1)%*%t(f1_obs) +matrix( X[ ,pos2], ncol=1)%*%t(f2_obs)             


plot(out$ind_fitted_func,  true_sig,
     xlab= "predicted value", ylab="true value") 
abline(a=0,b=1)
```

Note that these predictions can differ substantially from the observed noisy curves
```{r}

plot(out$ind_fitted_func,  Y,
     xlab= "predicted value", ylab="Noisy observation value") 
abline(a=0,b=1)
```


#### Handling function of any length and unevenly space data

Wavelets are primarily designed to analyze functions sampled on $2^J$ evenly spaced positions. However, this assumption is unrealistic in practice. For handling functions of any length, we use the approach of Kovac and Silvermann (2005), which basically remaps the data on a grid of length $2^J$. \textbf{What is important to know for the user:}  
* if you don't provide the sampling position (e.g., base pair measurement) in the pos argument, susiF considers that the columns of Y correspond to evenly spaced sampling positions. For example, when fine-mapping DNA methylation (DNAm) data, if you are trying to fine-map meQTL, the columns of the Y matrix correspond to different CpG . You \textbf{should} provide the genomic position of each of these CpG as a vector in the "pos" argument in the susiF function.


* susiF provides the estimated effect on the remapped effect on a $2^J$   grid. 


Example of how to analyze unevenly spaced data.
The black lines corresponds to  one of the SNP  effect in our data, the red ticks corresponds to the sampling positions and the red dots correspond to what part of the effect we can observe.

```{r echo=FALSE, message=FALSE, warning=FALSE}
set.seed(2)
data(N3finemapping)
attach(N3finemapping)

rsnr <- 0.5 #expected root signal noise ratio

lev_res <- 9#length of the molecular phenotype (2^lev_res)
f1 <-  simu_IBSS_per_level(lev_res )$sim_func  #first effect
f2 <- simu_IBSS_per_level(lev_res )$sim_func  #second effect


## Suppose we observe these functions on 200 points
sampling_pos <- sample( 1:length(f1), size=200)
sampling_pos <-sampling_pos  [order(sampling_pos)]

plot(f2, type="l", main="underlying function and the different sampling position")
 points(sampling_pos, f2[sampling_pos], col="red",pch=20)
 points(sampling_pos, rep(0, length(sampling_pos)), col="red",pch=3)
 abline(0,0)
 
```

Let's generate the data under these conditions


```{r message=FALSE, warning=FALSE}
  f1_obs <- f1[sampling_pos]
  f2_obs <- f2[sampling_pos]
noisy.data  <- list()
X <- N3finemapping$X[,1:100]
for ( i in 1:nrow(X))
{

  noise <- rnorm(length(f1_obs ), sd=  2)
  noisy.data [[i]] <-  X[i,pos1]*f1_obs +X[i,pos2]*f2_obs + noise

}
noisy.data <- do.call(rbind, noisy.data)

Y <- noisy.data

```


To account for it, pass the sampling position in the pos argument.



```{r message=FALSE, warning=FALSE}
 out <- susiF(Y,X,L=3 ,pos= sampling_pos)
```
To visualize the estimated effect with the correct x coordinate, use the "$outing_grid" object from the susiF.obj 

In this example, effect 2 (in our simulation) is captured in the first CS. The black line is the actual true effect. The red dots are the effect level at the sampling position, the thick blue line is the estimated effect, and the dashed line is the 95% credible bands.

```{r}
plot(   x=1:512,           y=f2,type="l")
points( x=sampling_pos,    y=f2_obs, col="red",pch=20)
lines(  x=out$outing_grid, y=out$fitted_func[[1]]    ,col='darkblue', lwd=2)
lines(  x=out$outing_grid, y= out$cred_band[[1]][1,] ,col='darkblue',lty=2 )
lines(  x=out$outing_grid, y= out$cred_band[[1]][2,] ,col='darkblue' ,lty=2 )
```
 
### A note on the priors available

Note that fSuSiE has two different priors available: \textit{mixture_normal_per_scale} and \textit{mixture_normal}. The default value is \textit{mixture_normal_per_scale}, which has slightly higher performance (power, estimation accuracy) than the  \textit{mixture_normal}. However,  \textit{mixture_normal} is up to 40% faster than the \textit{mixture_normal_per_scale}. You may consider using this option before performing genome-wide fine mapping. 
 
Here is a comparison between their running time
```{r }
 out1 <- susiF(Y,X,L=3 , prior = 'mixture_normal_per_scale',verbose=FALSE)
 out1$runtime
 out2 <- susiF(Y,X,L=3 , prior = 'mixture_normal',verbose=FALSE)
 out2$runtime
```
