---
title: "Introduction"
author: "William R.P. Denault"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{introduction}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.width = 5,
  fig.height = 3,
  fig.align = "center",
  fig.cap = "&nbsp;",
  dpi = 120
  )

 
```

# fSuSiE
 In this package, we implement a  Bayesian variable selection method called the "Functional Sum of Single Effects" ({\em fSuSiE}) model that extends the {\em SuSiE} model of Wang et al.^[this vignette is based on the original SuSiE vignette by [Gao Wang](https://stephenslab.github.io/susieR/articles/finemapping.html) ]  to functional phenotypes that can span over large genomic regions. The methods implemented here are particularly well-suited to settings where some of the X variables are highly correlated, and the true effects are highly sparse (e.g. <20 non-zero effects in the vector b). One example of this is genetic fine-mapping molecular measurements from sequencing assays (ChIP-seq, ATAC-seq ), as well as DNA methylation. However, the methods should also be useful more generally (e.g., detecting changes in time series).

###Model Description 
fSuSiE fits the following linear regression ${Y} = {XB} + {E}$, where ${Y}$ is a N x T matrix containing the observed curves, ${X}$ is the N × p matrix of single nucleotide polymorphisms (SNP) in the region to be fine-mapped. ${B}$ is a p × T matrix whose element $b_{jt}$ is the effect of SNP $j$ on the trait at location t, and ${E}$ is a matrix of error terms which initially, we consider being normally-distributed with variance $\sigma^2$. Analogous to regular fine-mapping, the matrix ${B}$ is expected to be sparse, i.e., most rows will be zero. Non-zero rows correspond to SNPs that do affect the trait. The goal is to identify the non-zero rows of  ${B}$. We extended  {\em SuSiE} to functional traits by writing ${B}= \sum_l^L {B}^{(l)}$, where each "single-effect" matrix ${B}^{(l)}$ has only one non-zero row, corresponding to a single causal SNP. We devised a fast variational inference procedure~\cite{blei_variational_2017} for fitting this model by first projecting the data in the wavelet space and then using an adaptation of the efficient iterative Bayesian stepwise selection (IBSS) approach proposed by Wang and colleagues.



The output of the fitting procedure is a number of "Credible Sets" (CSs), which are each designed to have a high probability of containing a variable with a non-zero effect, while at the same time being as small as possible. You can think of the CSs as being a set of "highly correlated" variables that are each associated with the response: you can be confident that one of the variables has a non-zero coefficient, but they are too correlated to be sure which one.

The package is developed by William R.P. Denault from the Stephens Lab at the University of Chicago.

Please post issues to ask questions, get our support or provide us feedback; please send pull requests if you have helped fix bugs or make improvements to the source code.


### Using fSuSiE

#### the data
This vignette shows how to use fsusieR in the context of genetic fine-mapping. We use simulated data of a molecular trait over a region of 124 base pairs (Y) with N≈600 individuals. We want to identify which columns of the genotype matrix X  (P=1000) cause changes in expression level.

The simulated data set is simulated to have exactly 2 non-zero effects.

Here we simulate the effects
```{r}
knitr::opts_chunk$set(
	echo = TRUE,
	message = FALSE,
	warning = FALSE
)
library(susiF.alpha)
library(susieR)
library(wavethresh)
set.seed(1)
data(N3finemapping)
attach(N3finemapping)

rsnr <- 0.5 #expected root signal noise ratio

pos1 <- 25    #Position of the causal covariate for effect 1
pos2 <- 75    #Position of the causal covariate for effect 2
lev_res <- 7#length of the molecular phenotype (2^lev_res)
f1 <-  simu_IBSS_per_level(lev_res )$sim_func#first effect
f2 <- simu_IBSS_per_level(lev_res )$sim_func #second effect

plot( f1, type ="l", ylab="effect", col="blue")
abline(a=0,b=0)
lines(f2, type="l", col="green")
legend(x=100,
       y=3,
       lty = rep(1,3),
      legend= c("effect 1", "effect 2" ),
       col=c( "blue","green"))

```
So the underlying causal variants are variant 25  and variant 75. The plot above shows the functional effect of the two variants. Below, we generate the  observed curves (effect +noise)



```{r}

noisy.data  <- list()
X <- N3finemapping$X[,1:100]
for ( i in 1:nrow(X))
{
  f1_obs <- f1
  f2_obs <- f2
  noise <- rnorm(length(f1), sd=  (1/  rsnr ) * var(f1))
  noisy.data [[i]] <-  X[i,pos1]*f1_obs +X[i,pos2]*f2_obs + noise

}
noisy.data <- do.call(rbind, noisy.data)

Y <- noisy.data


```

#### Output and interpretation


Running fSuSiE requires a Y matrix in which each line corresponds to an individual profile/time series. X the genotype matrix in which each line represents the individual genotype and the column the different SNPs. L is an upper bound on the number of causal SNPs within the regions (we advise keeping this parameter below 20).




```{r}
 out <- susiF(Y,X,L=10) 

```

FSuSiE automatically selects the number of causal effects (here two). The effect detected are summarized in terms of credible sets. Each credible set contains the likely causal covariates for a given effect. You can access them simply as 
 
```{r}
out$cs
```
As you can see the two CSs contain the correct causal SNPs.


### Visualization

The easiest way to visualize the result is to use the \code{plot_susiF} function

```{r}
 
 plot_susiF(out )
 

```





You can also access the information directly in the output of susiF in the fitted_func part of the output. See below 
 
 
```{r}
 par(mfrow=c(1,2))
 
 plot( f1, type="l", main="Estimated effect 1", xlab="")
 lines(get_fitted_effect(out,l=2),col='blue' )
 abline(a=0,b=0)
 legend(x= 35,
        y=3,
        lty= rep(1,2),
        legend = c("effect 1"," fSuSiE est "),
        col=c("black","blue" )
 )
 plot( f2, type="l", main="Estimated effect 2", xlab="")
 lines(get_fitted_effect(out,l=1),col='green' )
 abline(a=0,b=0)
 legend(x= 20,
        y=-1.5,
        lty= rep(1,2),
        legend = c("effect 2"," fSuSiE est "),
        col=c("black","green" )
 )
 
 par(mfrow=c(1,1))
```
 

####Posterior inclusion probabilities
You can also access the Posterior inclusion probabilities directly from the fitted bject. Similarly you can also specify the plot_susiF function only display the PiP plot.

 
```{r}
 
out$pip[1:10]

plot_susiF(out, pip_only=TRUE)
```
 
 
 
 
#### Extracting the regions affected by the credible sets
 We provide a build-in function to retreive the regions where the credible bands are "crossing zero"/i.e. the effects are likely not to be 0 in this region. 
```{r}
 affected_reg(susiF.obj = out)
  
```
 
 
#### A note on the priors available
Note that fSuSiE has two different priors available \textit{mixture_normal_per_scale} and \textit{mixture_normal}. The default value is \textit{mixture_normal_per_scale}, which has slightly higher performance (power, estimation accuracy) than the  \textit{mixture_normal}. However,  \textit{mixture_normal} is up to 40% faster than the \textit{mixture_normal_per_scale}. You may consider using this option before performing genome-wide fine mapping. 
 

```{r }
 out1 <- susiF(Y,X,L=3 , prior = 'mixture_normal_per_scale',verbose=FALSE)
 out1$runtime
 out2 <- susiF(Y,X,L=3 , prior = 'mixture_normal',verbose=FALSE)
 out2$runtime
```
 
 
#### Handling function of any length and unevely space data

Wavelets are primarly design to analyse function sampled on $2^J$  envely spaced position. However, this assumption is unrealistic in practice. For  handling function of any length we use the approach of Kovac and Silvermann (2005) that basically remaps the data on a grid of length $2^J$. \textbf{What is important to know for the user:}  
* if you don't provide the sampling position (e.g. base pair measurement) in the pos argument, susiF considers that the columns of Y corresponds to evenly spaced sampling positions. For exemple when fine mapping DNA methylation (DNAm) data, if you are trying to fine map meQTL, the columns of the Y matrix corresponds different CpG . You \textbf{should} provide the genomic position of each of this CpG as a vector in the "pos" argument in the susiF function.


* susiF provide the estimated effect on the remapped effect on the $2^J$ internal grid. 


Example of how to analysis unevenly spaced data.
The black lines corresponds to  one of the SNP  effect in our data, the red ticks corresponds to the sampling positions and the red dots correspond to what part of the effect we can observe.

```{r echo=FALSE, message=FALSE, warning=FALSE}
set.seed(2)
data(N3finemapping)
attach(N3finemapping)

rsnr <- 0.5 #expected root signal noise ratio

pos1 <- 25    #Position of the causal covariate for effect 1
pos2 <- 75    #Position of the causal covariate for effect 2
lev_res <- 9#length of the molecular phenotype (2^lev_res)
f1 <-  simu_IBSS_per_level(lev_res )$sim_func  #first effect
f2 <- simu_IBSS_per_level(lev_res )$sim_func  #second effect


## suppose we observe these functions on 200 points
sampling_pos <- sample( 1:length(f1), size=200)
sampling_pos <-sampling_pos  [order(sampling_pos)]

plot(f2, type="l", main="underlying function and the different sampling postion")
 points(sampling_pos, f2[sampling_pos], col="red",pch=20)
 points(sampling_pos, rep(0, length(sampling_pos)), col="red",pch=3)
 abline(0,0)
 
```

Lets generate the data under these conditions


```{r message=FALSE, warning=FALSE}
  f1_obs <- f1[sampling_pos]
  f2_obs <- f2[sampling_pos]
noisy.data  <- list()
X <- N3finemapping$X[,1:100]
for ( i in 1:nrow(X))
{

  noise <- rnorm(length(f1_obs ), sd=  2)
  noisy.data [[i]] <-  X[i,pos1]*f1_obs +X[i,pos2]*f2_obs + noise

}
noisy.data <- do.call(rbind, noisy.data)

Y <- noisy.data

```


To account for it simply pass the sampling position in the pos argument



```{r message=FALSE, warning=FALSE}
 out <- susiF(Y,X,L=3 ,pos= sampling_pos)
```
To visualize the estimated effect with the correct x coordinate use the "$outing_grid" object from the susiF.obj 

In this example effect 2 (in our simulation) is capture in the first CS. The black line the actual true effect. The red dots are the effect level at the sampling position, the thick blue line is the estimated effect and the dashed line are the 95% credible bands

```{r}
plot(   x=1:512,           y=f2,type="l")
points( x=sampling_pos,    y=f2_obs, col="red",pch=20)
lines(  x=out$outing_grid, y=out$fitted_func[[1]]    ,col='darkblue', lwd=2)
lines(  x=out$outing_grid, y= out$cred_band[[1]][1,] ,col='darkblue',lty=2 )
lines(  x=out$outing_grid, y= out$cred_band[[1]][2,] ,col='darkblue' ,lty=2 )
```




### Prediction

You can also check the predictive power of susiF by checking the fitted from the output susiF


Here the comparison betwe
```{r}
true_sig  <- matrix( X[ ,pos1], ncol=1)%*%t(f1_obs) +matrix( X[ ,pos2], ncol=1)%*%t(f2_obs)             


plot(out$ind_fitted_func,  true_sig) 
abline(a=0,b=1)
```

Note that these prediction can different substantially from the observed noisy curves
```{r}

plot(out$ind_fitted_func,  Y) 
abline(a=0,b=1)
```


