---
title: "fSuSiE: introduction to the key ideas with a toy methylation data set"
author: William Denault and Peter Carbonetto
date: "`r Sys.Date()`"
output:
  html_document:
    toc: no
    highlight: textmate
    theme: readable
vignette: >
  %\VignetteIndexEntry{}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

In this vignette we illustrate some of the key ideas underlying fSuSiE
by applying fSuSiE to analyze a toy methylation data set.

```{r knitr-opts, include=FALSE}
knitr::opts_chunk$set(comment = "#",collapse = TRUE,results = "hold",
                      fig.align = "center",dpi = 120)
```

Load the packages used in this vignette.

```{r load-pkgs}
library(fsusieR)
library(reshape2)
library(ggplot2)
library(cowplot)
```

Set the seed so that the results can be reproduced.

```{r set-seed}
set.seed(1)
```

Simulate data
-------------

We simulate a "toy" methylation data set in which the methylation
levels of 100 individuals are measured at 32 CpGs.

Among the 12 candidate SNPs, 3 SNPs affect the methylation levels: 2
SNPs affect the methylation levels of the same cluster of 8 CpGs, and
the other SNP affects the methylation levels of different cluster of 8
CpGs.

```{r sim-params}
n <- 100
m <- 32
p <- 12
```

Generate the SNP minor allele frequencies (MAFs):

```{r sim-mafs}
maf <- 0.05 + 0.45*runif(p)
```

Generate ids for SNPs and CpGs.

```{r sim-ids} 
snpids <- paste0("SNP-",1:p)
cpgids <- paste0("CpG-",1:m)
```

Simulate the genotypes:

```{r sim-geno}
X <- (runif(n*p) < maf) +
     (runif(n*p) < maf)
X <- matrix(X,n,p,byrow = TRUE)
storage.mode(X) <- "double"
X[,4] <- X[,3] + 0.03*rnorm(n)
colnames(X) <- snpids
```

This is the matrix that determines how the SNP alleles change the
methylation levels. 

```{r sim-effect-matrix}
F <- matrix(0,p,m)
F[1,9:16] <- 2.3
F[9,9:16] <- (-2.3)
F[3,25:32] <- 2
rownames(F) <- snpids
colnames(F) <- cpgids
```

Simulate the methylation levels at the CpGs:

```{r sim-Y}
E <- matrix(3*rnorm(n*m),n,m)
Y <- X %*% F
Y <- Y + E
baseline <- min(Y)
Y <- Y - baseline
```

To make the example more realistic, the methylation levels are all
simulated to be zero or higher.

To illustrate, this plot shows the effect of SNP 3 on the methylation
levels (the points show what the data look like):

```{r plot-data, fig.height=2, fig.width=5}
n1   <- 40
rows <- sample(n,n1)
pdat <- melt(Y[rows,])
x    <- X[rows,3]
pdat <- data.frame(cpg        = rep(1:m,each = n1) +
                                runif(m*n1,min = -0.2,max = 0.2),
                   meth_level = as.vector(Y[rows,]),
				   geno       = factor(x))
pdat_lines <- data.frame(cpg = rep(1:m,times = 3),
                         geno = factor(rep(0:2,each = m)),
	 			 		 meth_level = rep(c(rep(0,8),rep(-0.75,8),rep(0,16)),
 						                  times = 3))
pdat_lines$meth_level <- pdat_lines$meth_level - baseline
rows <- which(with(pdat_lines,geno == 1 & cpg > 24))
pdat_lines[rows,"meth_level"] <- pdat_lines[rows,"meth_level"] + 2
rows <- which(with(pdat_lines,geno == 2 & cpg > 24))
pdat_lines[rows,"meth_level"] <- pdat_lines[rows,"meth_level"] + 4
p1 <- ggplot(pdat,aes(x = cpg,y = meth_level,color = geno)) +
  geom_point(shape = 20,size = 1.25) +
  scale_color_manual(values = c("darkblue","limegreen","darkorange")) +
  geom_line(data = pdat_lines,size = 0.75) +
  labs(x = "CpG",y = "methylation level",color = "genotype") +
  theme_cowplot(font_size = 11)
print(p1)
```

SNP 3 affects CpGs 25–32 only.

```{r plot-data-ggsave, echo=FALSE, eval=TRUE}
ggsave("demo_data.pdf",
       p1 + scale_x_continuous(breaks = 1:m),
	   height = 2,width = 5)
```

QTL mapping
-----------

In a typical methylation QTL mapping analysis, one would perform
association tests for all the CpG-SNP pairs. Here we will simply use
the standard linear regression function in R, `lm()`, to perform the
QTL mapping.

```{r map-qtls}
assoc <- matrix(0,m,p)
rownames(assoc) <- cpgids
colnames(assoc) <- snpids
for (i in 1:m) {
  for (j in 1:p) {
    dat <- data.frame(x = X[,j],y = Y[,i])
    fit <- lm(y ~ x,dat)
    assoc[i,j] <- summary(fit)$coefficients["x","Pr(>|t|)"]
  }
}
```

Having performed these tests for association, we can examine the
associations in two different ways, by SNP and by CpG site. Let's
start with the SNP-centered view:

```{r gwas-by-snp, fig.height=2.5, fig.width=3}
pdat <- data.frame(cpg    = rep(1:m,times = p),
                   snp    = rep(1:p,each = m),
				   effect = as.vector(t(F != 0)),
                   pval   = as.vector(assoc))
pdat <- transform(pdat,pval = -log10(pval))
threshold <- -log10(0.05/(m*p))
p2 <- ggplot(pdat,aes(x = snp,y = pval,shape = effect,color = effect)) +
  geom_point(size = 1.5) +
  geom_hline(yintercept = threshold,color = "black",linetype = "dotted") +
  scale_x_continuous(breaks = 1:p) +
  scale_shape_manual(values = c(1,4)) +
  scale_color_manual(values = c("darkblue","darkorange")) +
  labs(x = "SNP",y = "-log10 p-value") +
  theme_cowplot(font_size = 11)
print(p2)
```

Based on these simple association tests, we would identify two out of
the three causal SNPs. (And perhaps we would identify all three causal
SNPs if we were a bit more careful about multiple testing
correction—here we just used the basic Bonferroni correction, which
tells us that only the CpG-SNPs pairs with p-values less than 0.0001
are significant.)

```{r gwas-by-snp-ggsave, echo=FALSE, eval=TRUE}
ggsave("demo_gwas_by_snp.pdf",p2,height = 3,width = 3)
```

*But more importantly*, this view alone doesn't tell us which CPGs are
affected by the SNPs.

This is the CpG-centered view:

```{r gwas-by-cpg, fig.height=2.5, fig.width=3.5}
p3 <- ggplot(pdat,aes(x = cpg,y = pval,shape = effect,color = effect)) +
  geom_point(size = 1.5) +
  geom_hline(yintercept = threshold,linetype = "dotted") +
  scale_x_continuous(breaks = c(0,seq(4,32,4))) +
  scale_shape_manual(values = c(1,4)) +
  scale_color_manual(values = c("darkblue","darkorange")) +
  labs(x = "CpG",y = "-log10 p-value") +
  theme_cowplot(font_size = 11)
print(p3)
```

From the association test p-values alone it is quite clear that the
affected CpG sites "cluster" in two continuous regions, but it is less
clear exactly which CpG sites in these clusters are affected.

```{r gwas-by-cpg-ggsave, echo=FALSE, eval=TRUE}
ggsave("demo_gwas_by_cpg.pdf",
       p3 + scale_x_continuous(breaks = seq(1,32)),
	   height = 3,width = 3.75)
```

*But more importantly*, this view alone doesn't tell us which SNPs are
affecting the changes in the methylation levels at these CpGs.

Fine-mapping with fSuSiE (with HMM post-processing step)
--------------------------------------------------------

Let's now contrast the above QTL mapping analysis with an analysis of
the methylation data using fSuSiE. First we fit the fSuSiE model to
the data. *Notice that we only a fit a single model to all the data*:

```{r run-fsusie, message=FALSE, results="hide"}
fit <- susiF(Y,X,L = 3,filter_cs = FALSE,prior = "mixture_normal",
             post_processing = "HMM")
```

(In this example it is assumed that we know that there are three
causal SNPs, but more generally this number can be estimated.)

fSuSiE indeed correctly produced 3 credible sets (95% CSs) each
containing one of the causal SNPs:

```{r fsusie-cs}
fit$cs
```

This result can be visualized using a "PIP plot" (PIP = posterior
inclusion probability):

```{r fsusie-pip-plot, fig.height=1.5, fig.width=3}
pdat <- data.frame(SNP = 1:p,
                   PIP = fit$pip,
				   CS  = rep("none",p))
pdat$CS[fit$cs[[1]]] <- "L1"
pdat$CS[fit$cs[[2]]] <- "L2"
pdat$CS[fit$cs[[3]]] <- "L3"
pdat <- transform(pdat,CS = factor(CS))
p4 <- ggplot(pdat,aes(x = SNP,y = PIP,fill = CS)) +
  geom_point(shape = 21,size = 2,color = "white") +
  scale_x_continuous(breaks = 1:p) +
  scale_fill_manual(values = c("dodgerblue","darkorange","red","gray")) +
  theme_cowplot(font_size = 10)
print(p4)
```

Again, fSuSiE very confidently identified the correct three causal
SNPs, whereas this was not the case in the QTL mapping. The key
difference is that fSuSiE fits a single model to all the data.

```{r fsusie-pip-plot-ggsave, echo=FALSE, eval=TRUE}
ggsave("demo_pip_plot.pdf",p4,height = 1.5,width = 3)
```

fSuSiE also gives a more coherent view of changes to the CpG sites and
how they are affected by the individual causal SNPs (or, more
precisely, by the individual CSs):

```{r fsusie-affected-cpgs, fig.height=2, fig.width=6, eval=TRUE}
i <- sapply(fit$cs,function (x) intersect(x,c(1,3,9)))
pdat <- data.frame(CS   = rep(c("L1","L2","L3"),each = m),
                   CpG  = rep(1:m,times = 3),
                   lfsr = unlist(fit$lfsr_func),
				   affected = as.vector(t(F[i,]) != 0),
				   stringsAsFactors = FALSE)
pdat <- transform(pdat,
                  CS   = factor(CS,c("L3","L2","L1")),
				  lfsr = -log10(lfsr))
p4 <- ggplot(pdat,aes(x = CpG,y = CS,size = lfsr,color = affected)) +
  geom_point(shape = 1) +
  scale_x_continuous(breaks = c(0,seq(4,32,4))) +
  scale_color_manual(values = c("darkblue","darkorange")) +
  scale_size(range = c(0.5,10),breaks = c(1.3,5,10)) +
  labs(size = "-log10(lfsr)") +
  theme_cowplot(font_size = 10)
print(p4)
```

Indeed, the significance tests (local false sign rate) break down the
support for affected CpGs *separately for each CS* (i.e., for each
causal SNP), and therefore it is quite evident from the fSuSiE results
that two of the causal SNPs affect the same CpG cluster, and the other
causal SNP affects a different CpG cluster.

```{r fsusie-affected-cpgs-ggsave, echo=FALSE, eval=TRUE}
ggsave("demo_affected_cpgs.pdf",
       p4 + scale_x_continuous(breaks = seq(1,32)),
	   height = 1.75,width = 6)
```

The *local false sign rates* (*lfsr*s) quantify the support the CpG
sites being affected, and in this example the CpG sites with the
smallest lfsrs (largest circles in the plot) are indeed the affected
CpGs.

Session info
------------

This is the version of R and the packages that were used to generate
these results.

```{r session-info}
sessionInfo()
```
