<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to fSuSiE • fsusieR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to fSuSiE">
<meta property="og:description" content="fsusieR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">fsusieR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.75</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/index.html">Vignettes</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/stephenslab/susiF.alpha" class="external-link">Source</a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introduction to fSuSiE</h1>
                        <h4 data-toc-skip class="author">William R.P.
Denault</h4>
            
            <h4 data-toc-skip class="date">2024-07-08</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/stephenslab/fsusieR/blob/HEAD/vignettes/fsusie_intro.Rmd" class="external-link"><code>vignettes/fsusie_intro.Rmd</code></a></small>
      <div class="hidden name"><code>fsusie_intro.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>fSuSiE implements a Bayesian variable selection method based on the
“Functional Sum of Single Effects” (fSuSiE) model. The fSuSiE model
extends the SuSiE model to functional traits. fSuSiE (and SuSiE) are
particularly well suited to settings where some of the variables are
strongly correlated and the true effects on the trait are very sparse.
This is exactly the situation we face in genetic fine-mapping. The
functional traits of interest are molecular measurements from sequencing
assays (e.g., ChIP-seq, ATAC-seq), and DNA methylation. However, the
methods are quite general and could be useful for other problems inside
and outside genetics.</p>
</div>
<div class="section level2">
<h2 id="the-fsusie-model">The fSuSiE model<a class="anchor" aria-label="anchor" href="#the-fsusie-model"></a>
</h2>
<p>fSuSiE fits the linear regression model <span class="math display">\[
{\bf Y} = {\bf X} {\bf B} + {\bf E},
\]</span></p>
<p>where <span class="math inline">\(\bf Y\)</span> is an N x T matrix
containing the <span class="math inline">\(N\)</span> trait observations
at <span class="math inline">\(T\)</span> locations; <span class="math inline">\(\bf X\)</span> is the N x p matrix of single
nucleotide polymorphisms (SNP) in the region to be fine-mapped; <span class="math inline">\(\bf B\)</span> is a p x T matrix whose element
<span class="math inline">\(b_{jt}\)</span> is the effect of SNP <span class="math inline">\(j\)</span> on the trait at location t; and <span class="math inline">\(\bf E\)</span> is a matrix of error terms.</p>
<p>Analogous to regular fine-mapping, the matrix <span class="math inline">\(\bf B\)</span> is expected to be row-wise sparse,
that is, most of the rows of <span class="math inline">\(\bf B\)</span>
will be zero. (Non-zero rows correspond to SNPs that affect the trait.)
The goal, in short, is to identify the non-zero rows of <span class="math inline">\({\bf B}\)</span>. To accomplish this task, we
extend SuSiE to functional traits: we express <span class="math inline">\({\bf B}\)</span> as a sum of “single effects”,
<span class="math display">\[
{\bf B} = \sum_{l=1}^L {B}^{(l)},
\]</span> where each single-effect matrix <span class="math inline">\({B}^{(l)}\)</span> has only one non-zero row,
corresponding to a single causal SNP. We have developed a fast procedure
for fitting this model by (i) projecting the data in the wavelet space,
then (ii) using a version of the iterative Bayesian stepwise selection
(IBSS) algorithm adapted for fSuSiE.</p>
<p>fSuSiE also outputs Credible Sets (CSs). Each CS is defined so that
it has a high probability of containing a variable with a non-zero
effect while at the same time being as small as possible.</p>
<p>We illustrate these ideas in the examples below.</p>
</div>
<div class="section level2">
<h2 id="the-example-fine-mapping-data-set">The example fine-mapping data set<a class="anchor" aria-label="anchor" href="#the-example-fine-mapping-data-set"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stephenslab/susieR" class="external-link">susieR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stephenslab/fsusieR" class="external-link">fsusieR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/" class="external-link">cowplot</a></span><span class="op">)</span></span></code></pre></div>
<p>We illustrate fSuSiE on a simulated molecular trait in which two of
the candidate genetic variants (SNPs) affect the trait. The molecular
trait spans a region of 120 locations.</p>
<p>First, we simulate the SNP effects (blue and green in the plot):</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">f1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simu_IBSS_per_level.html">simu_IBSS_per_level</a></span><span class="op">(</span><span class="fl">7</span><span class="op">)</span><span class="op">$</span><span class="va">sim_func</span></span>
<span><span class="va">f2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simu_IBSS_per_level.html">simu_IBSS_per_level</a></span><span class="op">(</span><span class="fl">7</span><span class="op">)</span><span class="op">$</span><span class="va">sim_func</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">2</span>,<span class="fl">2</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">f1</span>,type <span class="op">=</span> <span class="st">"l"</span>,col <span class="op">=</span> <span class="st">"royalblue"</span>,lwd <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>a <span class="op">=</span> <span class="fl">0</span>,b <span class="op">=</span> <span class="fl">0</span>,lty <span class="op">=</span> <span class="st">"dotted"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/lines.html" class="external-link">lines</a></span><span class="op">(</span><span class="va">f2</span>,type <span class="op">=</span> <span class="st">"l"</span>,col <span class="op">=</span> <span class="st">"limegreen"</span>,lwd <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="fsusie_intro_files/figure-html/sim-effects-1.png" alt="&amp;nbsp;" width="600"><p class="caption">
 
</p>
</div>
<p>Notice that each of the SNPs increases or decreases at many
locations, but in a mostly spatially smooth way.</p>
<p>Now we simulate the molecular trait. To do this, we use some genotype
data included in susieR.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">N3finemapping</span><span class="op">)</span></span>
<span><span class="va">rsnr</span>  <span class="op">&lt;-</span> <span class="fl">0.5</span></span>
<span><span class="va">pos1</span>  <span class="op">&lt;-</span> <span class="fl">25</span></span>
<span><span class="va">pos2</span>  <span class="op">&lt;-</span> <span class="fl">75</span></span>
<span><span class="va">X</span>     <span class="op">&lt;-</span> <span class="va">N3finemapping</span><span class="op">$</span><span class="va">X</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">]</span></span>
<span><span class="va">n</span>     <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span></span>
<span><span class="va">nloc</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">f1</span><span class="op">)</span></span>
<span><span class="va">noise</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="va">n</span><span class="op">*</span><span class="va">nloc</span>,sd <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">var</a></span><span class="op">(</span><span class="va">f1</span><span class="op">)</span><span class="op">/</span><span class="va">rsnr</span><span class="op">)</span>,<span class="va">n</span>,<span class="va">nloc</span><span class="op">)</span></span>
<span><span class="va">Y</span>     <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/crossprod.html" class="external-link">tcrossprod</a></span><span class="op">(</span><span class="va">X</span><span class="op">[</span>,<span class="va">pos1</span><span class="op">]</span>,<span class="va">f1</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/crossprod.html" class="external-link">tcrossprod</a></span><span class="op">(</span><span class="va">X</span><span class="op">[</span>,<span class="va">pos2</span><span class="op">]</span>,<span class="va">f2</span><span class="op">)</span> <span class="op">+</span> <span class="va">noise</span></span></code></pre></div>
<p>The end result is a fine-mapping data set containing the genotype
data at 100 SNPs (stored as a 574 x 100 matrix <span class="math inline">\({\bf X}\)</span>) and the molecular trait data at
128 locations (stored as a 574 x 128 matrix <span class="math inline">\({\bf Y}\)</span>).</p>
<p>Note that the true causal SNPs are at SNP positions 25 and 75
(columns 25 and 75 in the X matrix).</p>
</div>
<div class="section level2">
<h2 id="fitting-the-fsusie-model">Fitting the fSuSiE model<a class="anchor" aria-label="anchor" href="#fitting-the-fsusie-model"></a>
</h2>
<p>The main function in fsusieR for performing an fSuSiE analysis is the
“susiF” function. It has many input arguments for customizing or
fine-tuning the analysis, but at its simplest, only three inputs are
needed: (1) the molecular trait matrix, <span class="math inline">\(\bf
Y\)</span>; (2) the genotype matrix, <span class="math inline">\(\bf
X\)</span>; and (3) a number specifying an upper bound on the number of
single effects. Typically this number will be around 10, or perhaps a
bit larger. Therefore, the call to susiF for this simulated data set is
simply:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/susiF.html">susiF</a></span><span class="op">(</span><span class="va">Y</span>,<span class="va">X</span>,L <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<p>On most current computers, this call to susiF should not take more
than a minute or two to run.</p>
<p>fSuSiE automatically selects the number of causal effects (here two).
The effects detected are summarized in terms of credible sets (CSs).
Each credible set contains the likely causal covariates for a given
effect. You can access them simply as</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit</span><span class="op">$</span><span class="va">cs</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt; [1] 75 83</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[2]]</span></span>
<span><span class="co">#&gt; [1] 25</span></span></code></pre></div>
<p>As you can see, the two CSs contain the correct causal SNPs.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cs1</span> <span class="op">&lt;-</span> <span class="va">fit</span><span class="op">$</span><span class="va">cs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">cs2</span> <span class="op">&lt;-</span> <span class="va">fit</span><span class="op">$</span><span class="va">cs</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">fit</span><span class="op">$</span><span class="va">pip</span><span class="op">[</span><span class="va">cs1</span><span class="op">]</span></span>
<span><span class="co">#&gt; [1] 0.5 0.5</span></span>
<span><span class="va">fit</span><span class="op">$</span><span class="va">pip</span><span class="op">[</span><span class="va">cs2</span><span class="op">]</span></span>
<span><span class="co">#&gt; [1] 1</span></span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">X</span><span class="op">[</span>,<span class="va">cs1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;      [,1] [,2]</span></span>
<span><span class="co">#&gt; [1,]    1    1</span></span>
<span><span class="co">#&gt; [2,]    1    1</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="visualization">Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a>
</h2>
<p>You can also access the information directly in the output of susiF
in the fitted_func part of the output. See below</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/fsusie_plots.html">plot_susiF_effect</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="fsusie_intro_files/figure-html/plot-effects-1.png" alt="&amp;nbsp;" width="600"><p class="caption">
 
</p>
</div>
<p>You can also access the posterior inclusion probabilities directly
from the fitted object. Similarly, you can specify the plot_susiF
function to display only the PiP plot.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/fsusie_plots.html">plot_susiF_pip</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="fsusie_intro_files/figure-html/plot-pips-1.png" alt="&amp;nbsp;" width="600"><p class="caption">
 
</p>
</div>
<p>Or use the “plot” method to quickly generate both plots.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">pip</span>,<span class="va">res</span><span class="op">$</span><span class="va">effect</span>,</span>
<span>          nrow <span class="op">=</span> <span class="fl">2</span>,ncol <span class="op">=</span> <span class="fl">1</span>,</span>
<span>          rel_heights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>,<span class="fl">5</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="fsusie_intro_files/figure-html/plot-all-1.png" alt="&amp;nbsp;" width="600"><p class="caption">
 
</p>
</div>
</div>
<div class="section level2">
<h2 id="handling-functions-of-any-length-and-unevenly-spaced-data">Handling functions of any length and unevenly spaced data<a class="anchor" aria-label="anchor" href="#handling-functions-of-any-length-and-unevenly-spaced-data"></a>
</h2>
<p>Wavelets are primarily designed to analyze functions sampled on <span class="math inline">\(2^J\)</span> evenly spaced positions. However,
this assumption can be limiting.</p>
<p>For handling functions of any length, we use the approach of Kovac
and Silvermann (2005) which remaps the data on a grid of length <span class="math inline">\(2^J\)</span>.</p>
<p>Here are a couple of things to keep in mind:</p>
<ol style="list-style-type: decimal">
<li><p>If you don’t provide the sampling positions (e.g., base pair
measurement) in the “pos” argument, susiF assumes that the columns of Y
are evenly spaced. For molecular traits with known genomic positions
(e.g., base-pair positions), we recommend providing the genomic
positions in the “pos” argument.</p></li>
<li><p>susiF always returns the estimated effect on a <span class="math inline">\(2^J\)</span> grid.</p></li>
</ol>
<p>To illustrate, consider the following example. The true function is
defined on <span class="math inline">\(2^9 = 512\)</span> positions, but
we only observe the function in a random subset of those positions.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">N3finemapping</span><span class="op">)</span></span>
<span><span class="va">f1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simu_IBSS_per_level.html">simu_IBSS_per_level</a></span><span class="op">(</span><span class="fl">9</span><span class="op">)</span><span class="op">$</span><span class="va">sim_func</span></span>
<span><span class="va">f2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simu_IBSS_per_level.html">simu_IBSS_per_level</a></span><span class="op">(</span><span class="fl">9</span><span class="op">)</span><span class="op">$</span><span class="va">sim_func</span></span>
<span><span class="va">pos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">512</span>,<span class="fl">36</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">2</span>,<span class="fl">2</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">f2</span>,type <span class="op">=</span> <span class="st">"l"</span>,lwd <span class="op">=</span> <span class="fl">1.25</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/points.html" class="external-link">points</a></span><span class="op">(</span><span class="va">pos</span>,<span class="va">f2</span><span class="op">[</span><span class="va">pos</span><span class="op">]</span>,col <span class="op">=</span> <span class="st">"black"</span>,pch <span class="op">=</span> <span class="fl">20</span>,cex <span class="op">=</span> <span class="fl">0.75</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>a <span class="op">=</span> <span class="fl">0</span>,b <span class="op">=</span> <span class="fl">0</span>,col <span class="op">=</span> <span class="st">"black"</span>,lty <span class="op">=</span> <span class="st">"dotted"</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="fsusie_intro_files/figure-html/sim-effects-uneven-1.png" alt="&amp;nbsp;" width="600"><p class="caption">
 
</p>
</div>
<p>Let’s now generate the data matrix Y similar to above.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rsnr</span>   <span class="op">&lt;-</span> <span class="fl">0.03</span></span>
<span><span class="va">f1_obs</span> <span class="op">&lt;-</span> <span class="va">f1</span><span class="op">[</span><span class="va">pos</span><span class="op">]</span></span>
<span><span class="va">f2_obs</span> <span class="op">&lt;-</span> <span class="va">f2</span><span class="op">[</span><span class="va">pos</span><span class="op">]</span></span>
<span><span class="va">X</span>      <span class="op">&lt;-</span> <span class="va">N3finemapping</span><span class="op">$</span><span class="va">X</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">]</span></span>
<span><span class="va">n</span>      <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span></span>
<span><span class="va">nloc</span>   <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">pos</span><span class="op">)</span></span>
<span><span class="va">noise</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="va">n</span><span class="op">*</span><span class="va">nloc</span>,sd <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">var</a></span><span class="op">(</span><span class="va">f1</span><span class="op">)</span><span class="op">/</span><span class="va">rsnr</span><span class="op">)</span>,<span class="va">n</span>,<span class="va">nloc</span><span class="op">)</span></span>
<span><span class="va">Y</span>      <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/crossprod.html" class="external-link">tcrossprod</a></span><span class="op">(</span><span class="va">X</span><span class="op">[</span>,<span class="va">pos1</span><span class="op">]</span>,<span class="va">f1_obs</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/crossprod.html" class="external-link">tcrossprod</a></span><span class="op">(</span><span class="va">X</span><span class="op">[</span>,<span class="va">pos2</span><span class="op">]</span>,<span class="va">f2_obs</span><span class="op">)</span> <span class="op">+</span> <span class="va">noise</span></span></code></pre></div>
<p>The only change to the call to susiF is that we now also provide the
sampling positions:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/susiF.html">susiF</a></span><span class="op">(</span><span class="va">Y</span>,<span class="va">X</span>,L <span class="op">=</span> <span class="fl">10</span>,<span class="va">pos</span><span class="op">)</span></span>
<span><span class="co">#&gt; Response matrix dimensions not equal to nx 2^J </span></span>
<span><span class="co">#&gt;  or unevenly spaced data </span></span>
<span><span class="co">#&gt;  interpolation procedure used</span></span></code></pre></div>
<p>To visualize the estimated effects, again we can use the
“plot_susiF_effect” function:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/fsusie_plots.html">plot_susiF_effect</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="fsusie_intro_files/figure-html/plot-effects-uneven-1.png" alt="&amp;nbsp;" width="600"><p class="caption">
 
</p>
</div>
</div>
<div class="section level2">
<h2 id="prior-on-the-effects">Prior on the effects<a class="anchor" aria-label="anchor" href="#prior-on-the-effects"></a>
</h2>
<p>Note that fSuSiE has two different priors available:
mixture_normal_per_scale and mixture_normal. The default value is
mixture_normal_per_scale, which has slightly higher performance overall
(power, estimation accuracy) than mixture_normal. However,
mixture_normal is somewhat faster:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit_mnps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/susiF.html">susiF</a></span><span class="op">(</span><span class="va">Y</span>,<span class="va">X</span>,L <span class="op">=</span> <span class="fl">10</span>,prior <span class="op">=</span> <span class="st">"mixture_normal_per_scale"</span>,verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Fine mapping done, refining effect estimates using cylce spinning wavelet transform"</span></span>
<span><span class="va">fit_mn</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/susiF.html">susiF</a></span><span class="op">(</span><span class="va">Y</span>,<span class="va">X</span>,L <span class="op">=</span> <span class="fl">10</span>,prior <span class="op">=</span> <span class="st">"mixture_normal"</span>,verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Fine mapping done, refining effect estimates using cylce spinning wavelet transform"</span></span>
<span><span class="va">fit_mnps</span><span class="op">$</span><span class="va">runtime</span></span>
<span><span class="co">#&gt;    user  system elapsed </span></span>
<span><span class="co">#&gt;   2.045   0.045   2.113</span></span>
<span><span class="va">fit_mn</span><span class="op">$</span><span class="va">runtime</span></span>
<span><span class="co">#&gt;    user  system elapsed </span></span>
<span><span class="co">#&gt;   1.074   0.031   1.109</span></span></code></pre></div>
<p>Therefore, for larger data sets you may want to use the
<code>prior = "mixture_normal"</code> option.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by William R. P. Denault.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
