<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to fSuSiE • fsusieR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to fSuSiE">
<meta property="og:description" content="fsusieR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">fsusieR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/index.html">Vignettes</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/stephenslab/susiF.alpha" class="external-link">Source</a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introduction to fSuSiE</h1>
                        <h4 data-toc-skip class="author">William R.P.
Denault</h4>
            
            <h4 data-toc-skip class="date">2024-06-28</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/stephenslab/fsusieR/blob/HEAD/vignettes/introduction.Rmd" class="external-link"><code>vignettes/introduction.Rmd</code></a></small>
      <div class="hidden name"><code>introduction.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="fsusie">fSuSiE<a class="anchor" aria-label="anchor" href="#fsusie"></a>
</h2>
<p>In this package, we implement a Bayesian variable selection method
called the “Functional Sum of Single Effects” ({}) model that extends
the {} model of Wang et al.<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a> to functional phenotypes that can span over
large genomic regions. The methods implemented here are particularly
well-suited to settings where some of the X variables are highly
correlated, and the true effects are highly sparse (e.g., &lt;20
non-zero effects in the vector b). One example of this is genetic
fine-mapping molecular measurements from sequencing assays (ChIP-seq,
ATAC-seq ), as well as DNA methylation. However, the methods should also
be useful more generally (e.g., detecting changes in time series).</p>
<p>###Model Description</p>
<p>fSuSiE fits the following linear regression <span class="math inline">\({Y} = {XB} + {E}\)</span>, where <span class="math inline">\({Y}\)</span> is a N x T matrix containing the
observed curves, <span class="math inline">\({X}\)</span> is the N × p
matrix of single nucleotide polymorphisms (SNP) in the region to be
fine-mapped. <span class="math inline">\({B}\)</span> is a p × T matrix
whose element <span class="math inline">\(b_{jt}\)</span> is the effect
of SNP <span class="math inline">\(j\)</span> on the trait at location
t, and <span class="math inline">\({E}\)</span> is a matrix of error
terms which initially, we consider being normally-distributed with
variance <span class="math inline">\(\sigma^2\)</span>. Analogous to
regular fine-mapping, the matrix <span class="math inline">\({B}\)</span> is expected to be sparse, i.e., most
rows will be zero. Non-zero rows correspond to SNPs that do affect the
trait. The goal is to identify the non-zero rows of <span class="math inline">\({B}\)</span>. We extended {} to functional traits
by writing <span class="math inline">\({B}= \sum_l^L {B}^{(l)}\)</span>,
where each “single-effect” matrix <span class="math inline">\({B}^{(l)}\)</span> has only one non-zero row,
corresponding to a single causal SNP. We devised a fast variational
inference procedure~ for fitting this model by first projecting the data
in the wavelet space and then using an adaptation of the efficient
iterative Bayesian stepwise selection (IBSS) approach proposed by Wang
and colleagues.</p>
<p>The output of the fitting procedure is a number of “Credible Sets”
(CSs), which are each designed to have a high probability of containing
a variable with a non-zero effect while at the same time being as small
as possible. You can think of the CSs as being a set of “highly
correlated” variables that are each associated with the response: you
can be confident that one of the variables has a non-zero coefficient,
but they are too correlated to be sure which one.</p>
<p>The package is developed by William R.P. Denault from the Stephens
Lab at the University of Chicago.</p>
<p>Please post issues to ask questions, get our support, or provide us
feedback: please send pull requests if you have helped fix bugs or make
improvements to the source code.</p>
<div class="section level4">
<h4 id="using-fsusie">Using fSuSiE<a class="anchor" aria-label="anchor" href="#using-fsusie"></a>
</h4>
<div class="section level5">
<h5 id="the-data">the data<a class="anchor" aria-label="anchor" href="#the-data"></a>
</h5>
<p>This vignette shows how to use fsusieR in the context of genetic
fine-mapping. We use simulated data of a molecular trait over a region
of 124 base pairs (Y) with N≈600 individuals. We want to identify which
columns of the genotype matrix X (P=1000) cause changes in expression
level.</p>
<p>The simulated data set is simulated to have exactly two non-zero
effects.</p>
<p>Here, we simulate the effects</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/knitr/man/opts_chunk.html" class="external-link">opts_chunk</a></span><span class="op">$</span><span class="fu">set</span><span class="op">(</span></span>
<span>    echo <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    message <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    warning <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stephenslab/fsusieR" class="external-link">fsusieR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stephenslab/susieR" class="external-link">susieR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">wavethresh</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: MASS</span></span>
<span><span class="co">#&gt; WaveThresh: R wavelet software, release 4.7.2, installed</span></span>
<span><span class="co">#&gt; Copyright Guy Nason and others 1993-2022</span></span>
<span><span class="co">#&gt; Note: nlevels has been renamed to nlevelsWT</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">N3finemapping</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/attach.html" class="external-link">attach</a></span><span class="op">(</span><span class="va">N3finemapping</span><span class="op">)</span></span>
<span></span>
<span><span class="va">rsnr</span> <span class="op">&lt;-</span> <span class="fl">0.5</span> <span class="co">#expected root signal noise ratio</span></span>
<span></span>
<span><span class="va">pos1</span> <span class="op">&lt;-</span> <span class="fl">25</span>    <span class="co">#Position of the causal covariate for effect 1</span></span>
<span><span class="va">pos2</span> <span class="op">&lt;-</span> <span class="fl">75</span>    <span class="co">#Position of the causal covariate for effect 2</span></span>
<span><span class="va">lev_res</span> <span class="op">&lt;-</span> <span class="fl">7</span><span class="co">#length of the molecular phenotype (2^lev_res)</span></span>
<span><span class="va">f1</span> <span class="op">&lt;-</span>  <span class="fu"><a href="../reference/simu_IBSS_per_level.html">simu_IBSS_per_level</a></span><span class="op">(</span><span class="va">lev_res</span> <span class="op">)</span><span class="op">$</span><span class="va">sim_func</span><span class="co">#first effect</span></span>
<span><span class="va">f2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simu_IBSS_per_level.html">simu_IBSS_per_level</a></span><span class="op">(</span><span class="va">lev_res</span> <span class="op">)</span><span class="op">$</span><span class="va">sim_func</span> <span class="co">#second effect</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span> <span class="va">f1</span>, type <span class="op">=</span><span class="st">"l"</span>, ylab<span class="op">=</span><span class="st">"effect"</span>, col<span class="op">=</span><span class="st">"blue"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>a<span class="op">=</span><span class="fl">0</span>,b<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/lines.html" class="external-link">lines</a></span><span class="op">(</span><span class="va">f2</span>, type<span class="op">=</span><span class="st">"l"</span>, col<span class="op">=</span><span class="st">"green"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span>x<span class="op">=</span><span class="fl">100</span>,</span>
<span>       y<span class="op">=</span><span class="fl">3</span>,</span>
<span>       lty <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">3</span><span class="op">)</span>,</span>
<span>      legend<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"effect 1"</span>, <span class="st">"effect 2"</span> <span class="op">)</span>,</span>
<span>       col<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span> <span class="st">"blue"</span>,<span class="st">"green"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="introduction_files/figure-html/unnamed-chunk-2-1.png" alt="&amp;nbsp;" width="480"><p class="caption">
 
</p>
</div>
<p>So, the underlying causal variants are variant 25 and variant 75. The
plot above shows the functional effect of the two variants. Below, we
generate the observed curves (effect +noise)</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">noisy.data</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">X</span> <span class="op">&lt;-</span> <span class="va">N3finemapping</span><span class="op">$</span><span class="va">X</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">]</span></span>
<span><span class="kw">for</span> <span class="op">(</span> <span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">{</span></span>
<span>  <span class="va">f1_obs</span> <span class="op">&lt;-</span> <span class="va">f1</span></span>
<span>  <span class="va">f2_obs</span> <span class="op">&lt;-</span> <span class="va">f2</span></span>
<span>  <span class="va">noise</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">f1</span><span class="op">)</span>, sd<span class="op">=</span>  <span class="op">(</span><span class="fl">1</span><span class="op">/</span>  <span class="va">rsnr</span> <span class="op">)</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">var</a></span><span class="op">(</span><span class="va">f1</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">noisy.data</span> <span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span>  <span class="va">X</span><span class="op">[</span><span class="va">i</span>,<span class="va">pos1</span><span class="op">]</span><span class="op">*</span><span class="va">f1_obs</span> <span class="op">+</span><span class="va">X</span><span class="op">[</span><span class="va">i</span>,<span class="va">pos2</span><span class="op">]</span><span class="op">*</span><span class="va">f2_obs</span> <span class="op">+</span> <span class="va">noise</span></span>
<span></span>
<span><span class="op">}</span></span>
<span><span class="va">noisy.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="va">noisy.data</span><span class="op">)</span></span>
<span></span>
<span><span class="va">Y</span> <span class="op">&lt;-</span> <span class="va">noisy.data</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="output-and-interpretation">Output and interpretation<a class="anchor" aria-label="anchor" href="#output-and-interpretation"></a>
</h5>
<p>Running fSuSiE requires a Y matrix in which each line corresponds to
an individual profile/time series. X the genotype matrix in which each
line represents the individual genotype and the column the different
SNPs. L is an upper bound on the number of causal SNPs within the
regions (we advise keeping this parameter below 20).</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span> <span class="va">out</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/susiF.html">susiF</a></span><span class="op">(</span><span class="va">Y</span>,<span class="va">X</span>,L<span class="op">=</span><span class="fl">10</span><span class="op">)</span> </span>
<span><span class="co">#&gt; [1] "Starting initialization"</span></span>
<span><span class="co">#&gt; [1] "Data transform"</span></span>
<span><span class="co">#&gt; [1] "Discarding  0 wavelet coefficients out of  128"</span></span>
<span><span class="co">#&gt; [1] "Data transform done"</span></span>
<span><span class="co">#&gt; [1] "Initializing prior"</span></span>
<span><span class="co">#&gt; [1] "Initialization done"</span></span>
<span><span class="co">#&gt; [1] "Fitting effect  1 , iter 1"</span></span>
<span><span class="co">#&gt; [1] "Fitting effect  2 , iter 1"</span></span>
<span><span class="co">#&gt; [1] "Fitting effect  3 , iter 1"</span></span>
<span><span class="co">#&gt; [1] "Discarding  1  effects"</span></span>
<span><span class="co">#&gt; [1] "Fitting effect  1 , iter 2"</span></span>
<span><span class="co">#&gt; [1] "Fitting effect  2 , iter 2"</span></span>
<span><span class="co">#&gt; [1] "Discarding  8  effects"</span></span>
<span><span class="co">#&gt; [1] "Greedy search and backfitting done"</span></span>
<span><span class="co">#&gt; [1] "Fitting effect  1 , iter 3"</span></span>
<span><span class="co">#&gt; [1] "Fitting effect  2 , iter 3"</span></span>
<span><span class="co">#&gt; [1] "Fine mapping done, refining effect estimates using cylce spinning wavelet transform"</span></span></code></pre></div>
<p>FSuSiE automatically selects the number of causal effects (here two).
The effects detected are summarized in terms of credible sets. Each
credible set contains the likely causal covariates for a given effect.
You can access them simply as</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">out</span><span class="op">$</span><span class="va">cs</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt; [1] 75 83</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[2]]</span></span>
<span><span class="co">#&gt; [1] 25</span></span></code></pre></div>
<p>As you can see, the two CSs contain the correct causal SNPs.</p>
</div>
</div>
<div class="section level4">
<h4 id="visualization">Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a>
</h4>
<p>The easiest way to visualize the result is to use the function.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span> </span>
<span> <span class="fu"><a href="../reference/plot_susiF.html">plot_susiF</a></span><span class="op">(</span><span class="va">out</span> <span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="introduction_files/figure-html/unnamed-chunk-6-1.png" alt="&amp;nbsp;" width="480"><p class="caption">
 
</p>
</div>
<div class="figure" style="text-align: center">
<img src="introduction_files/figure-html/unnamed-chunk-6-2.png" alt="&amp;nbsp;" width="480"><p class="caption">
 
</p>
</div>
<p>You can also access the information directly in the output of susiF
in the fitted_func part of the output. See below</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span> <span class="va">f1</span>, type<span class="op">=</span><span class="st">"l"</span>, main<span class="op">=</span><span class="st">"Estimated effect 1"</span>, xlab<span class="op">=</span><span class="st">""</span><span class="op">)</span></span>
<span> <span class="fu"><a href="https://rdrr.io/r/graphics/lines.html" class="external-link">lines</a></span><span class="op">(</span><span class="fu"><a href="../reference/get_fitted_effect.html">get_fitted_effect</a></span><span class="op">(</span><span class="va">out</span>,l<span class="op">=</span><span class="fl">2</span><span class="op">)</span>,col<span class="op">=</span><span class="st">'blue'</span> <span class="op">)</span></span>
<span> <span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>a<span class="op">=</span><span class="fl">0</span>,b<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span>
<span> <span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span>x<span class="op">=</span> <span class="fl">35</span>,</span>
<span>        y<span class="op">=</span><span class="fl">3</span>,</span>
<span>        lty<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>,</span>
<span>        legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"effect 1"</span>,<span class="st">" fSuSiE est "</span><span class="op">)</span>,</span>
<span>        col<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>,<span class="st">"blue"</span> <span class="op">)</span></span>
<span> <span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="introduction_files/figure-html/unnamed-chunk-7-1.png" alt="&amp;nbsp;" width="480"><p class="caption">
 
</p>
</div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span> <span class="va">f2</span>, type<span class="op">=</span><span class="st">"l"</span>, main<span class="op">=</span><span class="st">"Estimated effect 2"</span>, xlab<span class="op">=</span><span class="st">""</span><span class="op">)</span></span>
<span> <span class="fu"><a href="https://rdrr.io/r/graphics/lines.html" class="external-link">lines</a></span><span class="op">(</span><span class="fu"><a href="../reference/get_fitted_effect.html">get_fitted_effect</a></span><span class="op">(</span><span class="va">out</span>,l<span class="op">=</span><span class="fl">1</span><span class="op">)</span>,col<span class="op">=</span><span class="st">'green'</span> <span class="op">)</span></span>
<span> <span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>a<span class="op">=</span><span class="fl">0</span>,b<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span>
<span> <span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span>x<span class="op">=</span> <span class="fl">20</span>,</span>
<span>        y<span class="op">=</span><span class="op">-</span><span class="fl">1.5</span>,</span>
<span>        lty<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>,</span>
<span>        legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"effect 2"</span>,<span class="st">" fSuSiE est "</span><span class="op">)</span>,</span>
<span>        col<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>,<span class="st">"green"</span> <span class="op">)</span></span>
<span> <span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="introduction_files/figure-html/unnamed-chunk-7-2.png" alt="&amp;nbsp;" width="480"><p class="caption">
 
</p>
</div>
<p>####Posterior inclusion probabilities You can also access the
Posterior inclusion probabilities directly from the fitted object.
Similarly, you can specify the plot_susiF function to display only the
PiP plot.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span> </span>
<span><span class="va">out</span><span class="op">$</span><span class="va">pip</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span>
<span><span class="co">#&gt;  [1] 0 0 0 0 0 0 0 0 0 0</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot_susiF.html">plot_susiF</a></span><span class="op">(</span><span class="va">out</span>, pip_only<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="introduction_files/figure-html/unnamed-chunk-8-1.png" alt="&amp;nbsp;" width="480"><p class="caption">
 
</p>
</div>
<div class="section level5">
<h5 id="extracting-the-regions-affected-by-the-credible-sets">Extracting the regions affected by the credible sets<a class="anchor" aria-label="anchor" href="#extracting-the-regions-affected-by-the-credible-sets"></a>
</h5>
<p>We provide a build-in function to retrieve the regions where the
credible bands are “crossing zero”/i.e., the effects are likely not to
be 0 in this region.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span> <span class="fu"><a href="../reference/affected_reg.html">affected_reg</a></span><span class="op">(</span> <span class="va">out</span><span class="op">)</span></span>
<span><span class="co">#&gt;    CS Start End</span></span>
<span><span class="co">#&gt; 1   1     1   8</span></span>
<span><span class="co">#&gt; 2   1    12  16</span></span>
<span><span class="co">#&gt; 8   1    19  19</span></span>
<span><span class="co">#&gt; 9   1    21  26</span></span>
<span><span class="co">#&gt; 3   1    32  32</span></span>
<span><span class="co">#&gt; 4   1    34  34</span></span>
<span><span class="co">#&gt; 10  1    38  50</span></span>
<span><span class="co">#&gt; 11  1    59  78</span></span>
<span><span class="co">#&gt; 5   1    86  98</span></span>
<span><span class="co">#&gt; 6   1   104 120</span></span>
<span><span class="co">#&gt; 7   1   122 128</span></span>
<span><span class="co">#&gt; 17  2     1  16</span></span>
<span><span class="co">#&gt; 12  2    19  21</span></span>
<span><span class="co">#&gt; 13  2    23  24</span></span>
<span><span class="co">#&gt; 18  2    28  39</span></span>
<span><span class="co">#&gt; 14  2    44  47</span></span>
<span><span class="co">#&gt; 15  2    50  76</span></span>
<span><span class="co">#&gt; 16  2    86  90</span></span>
<span><span class="co">#&gt; 19  2   106 107</span></span>
<span><span class="co">#&gt; 20  2   120 128</span></span></code></pre></div>
</div>
</div>
<div class="section level4">
<h4 id="prediction">Prediction<a class="anchor" aria-label="anchor" href="#prediction"></a>
</h4>
<p>You can also check the predictive power of susiF by checking the
fitted from the output susiF</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">true_sig</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span> <span class="va">X</span><span class="op">[</span> ,<span class="va">pos1</span><span class="op">]</span>, ncol<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">f1_obs</span><span class="op">)</span> <span class="op">+</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span> <span class="va">X</span><span class="op">[</span> ,<span class="va">pos2</span><span class="op">]</span>, ncol<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">f2_obs</span><span class="op">)</span>             </span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">out</span><span class="op">$</span><span class="va">ind_fitted_func</span>,  <span class="va">true_sig</span>,</span>
<span>     xlab<span class="op">=</span> <span class="st">"predicted value"</span>, ylab<span class="op">=</span><span class="st">"true value"</span><span class="op">)</span> </span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>a<span class="op">=</span><span class="fl">0</span>,b<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="introduction_files/figure-html/unnamed-chunk-10-1.png" alt="&amp;nbsp;" width="480"><p class="caption">
 
</p>
</div>
<p>Note that these predictions can differ substantially from the
observed noisy curves</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">out</span><span class="op">$</span><span class="va">ind_fitted_func</span>,  <span class="va">Y</span>,</span>
<span>     xlab<span class="op">=</span> <span class="st">"predicted value"</span>, ylab<span class="op">=</span><span class="st">"Noisy observation value"</span><span class="op">)</span> </span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>a<span class="op">=</span><span class="fl">0</span>,b<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="introduction_files/figure-html/unnamed-chunk-11-1.png" alt="&amp;nbsp;" width="480"><p class="caption">
 
</p>
</div>
<div class="section level5">
<h5 id="handling-function-of-any-length-and-unevenly-space-data">Handling function of any length and unevenly space data<a class="anchor" aria-label="anchor" href="#handling-function-of-any-length-and-unevenly-space-data"></a>
</h5>
<p>Wavelets are primarily designed to analyze functions sampled on <span class="math inline">\(2^J\)</span> evenly spaced positions. However,
this assumption is unrealistic in practice. For handling functions of
any length, we use the approach of Kovac and Silvermann (2005), which
basically remaps the data on a grid of length <span class="math inline">\(2^J\)</span>. <br>
* if you don’t provide the sampling position (e.g., base pair
measurement) in the pos argument, susiF considers that the columns of Y
correspond to evenly spaced sampling positions. For example, when
fine-mapping DNA methylation (DNAm) data, if you are trying to fine-map
meQTL, the columns of the Y matrix correspond to different CpG . You
provide the genomic position of each of these CpG as a vector in the
“pos” argument in the susiF function.</p>
<ul>
<li>susiF provides the estimated effect on the remapped effect on a
<span class="math inline">\(2^J\)</span> grid.</li>
</ul>
<p>Example of how to analyze unevenly spaced data. The black lines
corresponds to one of the SNP effect in our data, the red ticks
corresponds to the sampling positions and the red dots correspond to
what part of the effect we can observe.</p>
<div class="figure" style="text-align: center">
<img src="introduction_files/figure-html/unnamed-chunk-12-1.png" alt="&amp;nbsp;" width="480"><p class="caption">
 
</p>
</div>
<p>Let’s generate the data under these conditions</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="va">f1_obs</span> <span class="op">&lt;-</span> <span class="va">f1</span><span class="op">[</span><span class="va">sampling_pos</span><span class="op">]</span></span>
<span>  <span class="va">f2_obs</span> <span class="op">&lt;-</span> <span class="va">f2</span><span class="op">[</span><span class="va">sampling_pos</span><span class="op">]</span></span>
<span><span class="va">noisy.data</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">X</span> <span class="op">&lt;-</span> <span class="va">N3finemapping</span><span class="op">$</span><span class="va">X</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">]</span></span>
<span><span class="kw">for</span> <span class="op">(</span> <span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">{</span></span>
<span></span>
<span>  <span class="va">noise</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">f1_obs</span> <span class="op">)</span>, sd<span class="op">=</span>  <span class="fl">2</span><span class="op">)</span></span>
<span>  <span class="va">noisy.data</span> <span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span>  <span class="va">X</span><span class="op">[</span><span class="va">i</span>,<span class="va">pos1</span><span class="op">]</span><span class="op">*</span><span class="va">f1_obs</span> <span class="op">+</span><span class="va">X</span><span class="op">[</span><span class="va">i</span>,<span class="va">pos2</span><span class="op">]</span><span class="op">*</span><span class="va">f2_obs</span> <span class="op">+</span> <span class="va">noise</span></span>
<span></span>
<span><span class="op">}</span></span>
<span><span class="va">noisy.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="va">noisy.data</span><span class="op">)</span></span>
<span></span>
<span><span class="va">Y</span> <span class="op">&lt;-</span> <span class="va">noisy.data</span></span></code></pre></div>
<p>To account for it, pass the sampling position in the pos
argument.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span> <span class="va">out</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/susiF.html">susiF</a></span><span class="op">(</span><span class="va">Y</span>,<span class="va">X</span>,L<span class="op">=</span><span class="fl">3</span> ,pos<span class="op">=</span> <span class="va">sampling_pos</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Starting initialization"</span></span>
<span><span class="co">#&gt; [1] "Data transform"</span></span>
<span><span class="co">#&gt; [1] "Discarding  0 wavelet coefficients out of  256"</span></span>
<span><span class="co">#&gt; [1] "Data transform done"</span></span>
<span><span class="co">#&gt; [1] "Initializing prior"</span></span>
<span><span class="co">#&gt; [1] "Initialization done"</span></span>
<span><span class="co">#&gt; [1] "Fitting effect  1 , iter 1"</span></span>
<span><span class="co">#&gt; [1] "Fitting effect  2 , iter 1"</span></span>
<span><span class="co">#&gt; [1] "Fitting effect  3 , iter 1"</span></span>
<span><span class="co">#&gt; [1] "Discarding  1  effects"</span></span>
<span><span class="co">#&gt; [1] "Fitting effect  1 , iter 2"</span></span>
<span><span class="co">#&gt; [1] "Fitting effect  2 , iter 2"</span></span>
<span><span class="co">#&gt; [1] "Discarding  1  effects"</span></span>
<span><span class="co">#&gt; [1] "Greedy search and backfitting done"</span></span>
<span><span class="co">#&gt; [1] "Fitting effect  1 , iter 3"</span></span>
<span><span class="co">#&gt; [1] "Fitting effect  2 , iter 3"</span></span>
<span><span class="co">#&gt; [1] "Fine mapping done, refining effect estimates using cylce spinning wavelet transform"</span></span></code></pre></div>
<p>To visualize the estimated effect with the correct x coordinate, use
the “$outing_grid” object from the susiF.obj</p>
<p>In this example, effect 2 (in our simulation) is captured in the
first CS. The black line is the actual true effect. The red dots are the
effect level at the sampling position, the thick blue line is the
estimated effect, and the dashed line is the 95% credible bands.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span>   x<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fl">512</span>,           y<span class="op">=</span><span class="va">f2</span>,type<span class="op">=</span><span class="st">"l"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/points.html" class="external-link">points</a></span><span class="op">(</span> x<span class="op">=</span><span class="va">sampling_pos</span>,    y<span class="op">=</span><span class="va">f2_obs</span>, col<span class="op">=</span><span class="st">"red"</span>,pch<span class="op">=</span><span class="fl">20</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/lines.html" class="external-link">lines</a></span><span class="op">(</span>  x<span class="op">=</span><span class="va">out</span><span class="op">$</span><span class="va">outing_grid</span>, y<span class="op">=</span><span class="va">out</span><span class="op">$</span><span class="va">fitted_func</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>    ,col<span class="op">=</span><span class="st">'darkblue'</span>, lwd<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/lines.html" class="external-link">lines</a></span><span class="op">(</span>  x<span class="op">=</span><span class="va">out</span><span class="op">$</span><span class="va">outing_grid</span>, y<span class="op">=</span> <span class="va">out</span><span class="op">$</span><span class="va">cred_band</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span>,<span class="op">]</span> ,col<span class="op">=</span><span class="st">'darkblue'</span>,lty<span class="op">=</span><span class="fl">2</span> <span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/lines.html" class="external-link">lines</a></span><span class="op">(</span>  x<span class="op">=</span><span class="va">out</span><span class="op">$</span><span class="va">outing_grid</span>, y<span class="op">=</span> <span class="va">out</span><span class="op">$</span><span class="va">cred_band</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">2</span>,<span class="op">]</span> ,col<span class="op">=</span><span class="st">'darkblue'</span> ,lty<span class="op">=</span><span class="fl">2</span> <span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="introduction_files/figure-html/unnamed-chunk-15-1.png" alt="&amp;nbsp;" width="480"><p class="caption">
 
</p>
</div>
</div>
<div class="section level5">
<h5 id="a-note-on-the-priors-available">A note on the priors available<a class="anchor" aria-label="anchor" href="#a-note-on-the-priors-available"></a>
</h5>
<p>Note that fSuSiE has two different priors available: and . The
default value is , which has slightly higher performance (power,
estimation accuracy) than the . However, is up to 40% faster than the .
You may consider using this option before performing genome-wide fine
mapping.</p>
<p>Here is a comparison between their running time</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span> <span class="va">out1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/susiF.html">susiF</a></span><span class="op">(</span><span class="va">Y</span>,<span class="va">X</span>,L<span class="op">=</span><span class="fl">3</span> , prior <span class="op">=</span> <span class="st">'mixture_normal_per_scale'</span>,verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Fine mapping done, refining effect estimates using cylce spinning wavelet transform"</span></span>
<span> <span class="va">out1</span><span class="op">$</span><span class="va">runtime</span></span>
<span><span class="co">#&gt;    user  system elapsed </span></span>
<span><span class="co">#&gt;   6.285   0.317   6.641</span></span>
<span> <span class="va">out2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/susiF.html">susiF</a></span><span class="op">(</span><span class="va">Y</span>,<span class="va">X</span>,L<span class="op">=</span><span class="fl">3</span> , prior <span class="op">=</span> <span class="st">'mixture_normal'</span>,verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Fine mapping done, refining effect estimates using cylce spinning wavelet transform"</span></span>
<span> <span class="va">out2</span><span class="op">$</span><span class="va">runtime</span></span>
<span><span class="co">#&gt;    user  system elapsed </span></span>
<span><span class="co">#&gt;   2.377   0.123   2.515</span></span></code></pre></div>
</div>
</div>
</div>
<div class="footnotes footnotes-end-of-document">
<hr>
<ol>
<li id="fn1"><p>this vignette is based on the original SuSiE vignette by
<a href="https://stephenslab.github.io/susieR/articles/finemapping.html" class="external-link">Gao
Wang</a><a href="#fnref1" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by William R. P. Denault.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
